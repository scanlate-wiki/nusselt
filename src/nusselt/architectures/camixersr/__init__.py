
import math
import re

from .camixersr import camixersr
from ...registry.model_descriptor import StateDict, ModelDescriptor

def load(state_dict: StateDict) -> ModelDescriptor[camixersr]:
    state_dict_keys = state_dict.keys()
    layer_dict = {}
    for r in state_dict_keys:
        g = re.match(r"body.(\d*).body.(\d*).norm1", r)
        if g:
            num_layer, num_block = g.groups()
            layer_dict[int(num_layer) + 1] = int(num_block) + 1

    n_group = len(layer_dict)
    n_block = [i for i in layer_dict.values()]
    ratio = 0.5  # I don't know if it's even possible to get it
    n_feats = state_dict["head.weight"].shape[0]
    model_tail = state_dict["tail.0.weight"].shape
    n_colors = model_tail[3]
    scale = int(math.sqrt(model_tail[0] / n_colors))

    model = camixersr(
        scale=scale,
        n_feats=n_feats,
        n_colors=n_colors,
        n_block=n_block,
        n_group=n_group,
        ratio=ratio

    )

    return ModelDescriptor(
        model,
        state_dict,
        architecture="camixersr",
        scale=scale,
        input_channels=n_colors,
        output_channels=n_colors
    )
