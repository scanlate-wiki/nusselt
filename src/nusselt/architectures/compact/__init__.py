from .SRVGG import SRVGGNetCompact as Compact
from ...registry.model_descriptor import StateDict, ModelDescriptor
from ...utilities.state import get_seq_len, get_scale_and_output_channels


def load(state_dict: StateDict) -> ModelDescriptor[Compact]:
    state = state_dict

    highest_num = get_seq_len(state, "body") - 1

    in_nc = state["body.0.weight"].shape[1]
    num_feat = state["body.0.weight"].shape[0]
    num_conv = (highest_num - 2) // 2

    pixelshuffle_shape = state[f"body.{highest_num}.bias"].shape[0]
    scale, out_nc = get_scale_and_output_channels(pixelshuffle_shape, in_nc)

    model = Compact(
        num_in_ch=in_nc,
        num_out_ch=out_nc,
        num_feat=num_feat,
        num_conv=num_conv,
        upscale=scale,
    )

    return ModelDescriptor(
        model,
        state_dict,
        architecture="Compact",
        scale=scale,
        input_channels=in_nc,
        output_channels=out_nc
    )
