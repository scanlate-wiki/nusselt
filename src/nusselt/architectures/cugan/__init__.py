from .cugan import cugan
from ...registry.model_descriptor import StateDict, ModelDescriptor

def load(state_dict: StateDict) -> ModelDescriptor[cugan]:
    if "conv_final.weight" in state_dict:
        # UpCunet4x
        scale = 4
        in_channels = state_dict["unet1.conv1.conv.0.weight"].shape[1]
        out_channels = 3  # hard coded in UpCunet4x
    elif state_dict["unet1.conv_bottom.weight"].shape[2] == 5:
        # UpCunet3x
        scale = 3
        in_channels = state_dict["unet1.conv1.conv.0.weight"].shape[1]
        out_channels = state_dict["unet2.conv_bottom.weight"].shape[0]
    else:
        # UpCunet2x
        scale = 2
        in_channels = state_dict["unet1.conv1.conv.0.weight"].shape[1]
        out_channels = state_dict["unet2.conv_bottom.weight"].shape[0]
    pro = False
    if list(state_dict.keys())[-1] == "unet2.conv_bottom.bias":
        pro = True
    if "pro" in state_dict:
        pro = True
        del state_dict["pro"]
    model = cugan(
        in_channels = in_channels,
        out_channels = out_channels,
        scale=scale,
        pro=pro

    )
    return ModelDescriptor(
        model,
        state_dict,
        architecture="cugan",
        scale=scale,
        input_channels=in_channels,
        output_channels=out_channels
    )
