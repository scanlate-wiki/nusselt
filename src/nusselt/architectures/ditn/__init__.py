from .ditn_real import DITN_Real
from ...registry.model_descriptor import ModelDescriptor, StateDict
from ...utilities.state import get_seq_len, get_scale_and_output_channels


def load(state_dict: StateDict) -> ModelDescriptor[DITN_Real]:
    LayerNorm_type = "WithBias"  # unused internally
    patch_size = 8  # cannot be deduced from state_dict

    inp_channels = state_dict["sft.weight"].shape[1]
    dim = state_dict["sft.weight"].shape[0]

    UFONE_blocks = get_seq_len(state_dict, "UFONE")
    ITL_blocks = get_seq_len(state_dict, "UFONE.0.ITLs")
    SAL_blocks = get_seq_len(state_dict, "UFONE.0.SALs")

    ffn_expansion_factor = state_dict["UFONE.0.ITLs.0.ffn.project_in.weight"].shape[0] / 2 / dim

    bias = "UFONE.0.ITLs.0.attn.project_out.bias" in state_dict

    upscale, _ = get_scale_and_output_channels(state_dict["upsample.0.weight"].shape[0], inp_channels)

    model = DITN_Real(
        inp_channels=inp_channels,
        dim=dim,
        ITL_blocks=ITL_blocks,
        SAL_blocks=SAL_blocks,
        UFONE_blocks=UFONE_blocks,
        ffn_expansion_factor=ffn_expansion_factor,
        bias=bias,
        LayerNorm_type=LayerNorm_type,
        patch_size=patch_size,
        upscale=upscale,
    )

    return ModelDescriptor(
        model=model,
        state_dict=state_dict,
        architecture="DITN",
        scale=upscale,
        input_channels=inp_channels,
        output_channels=3,
    )
