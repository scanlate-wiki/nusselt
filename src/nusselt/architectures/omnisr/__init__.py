import math

from .omnisr import OmniSR
from ...registry.model_descriptor import StateDict, ModelDescriptor
from ...utilities.state import get_scale_and_output_channels, get_seq_len


def load(state_dict: StateDict) -> ModelDescriptor[OmniSR]:
    window_size = 8

    num_feat = state_dict["input.weight"].shape[0]
    num_in_ch = state_dict["input.weight"].shape[1]

    upscale, num_out_ch = get_scale_and_output_channels(
        state_dict["up.0.weight"].shape[0],
        num_in_ch,
    )

    bias = "input.bias" in state_dict
    res_num = get_seq_len(state_dict, "residual_layer")
    block_num = get_seq_len(state_dict, "residual_layer.0.residual_layer") - 1

    rel_pos_bias_key = "residual_layer.0.residual_layer.0.layer.2.fn.rel_pos_bias.weight"
    if rel_pos_bias_key in state_dict:
        pe = True
        rel_pos_bias_weight = state_dict[rel_pos_bias_key].shape[0]
        window_size = int((math.sqrt(rel_pos_bias_weight) + 1) / 2)
    else:
        pe = False

    model = OmniSR(
        num_in_ch=num_in_ch,
        num_out_ch=num_out_ch,
        num_feat=num_feat,
        up_scale=upscale,
        res_num=res_num,
        pe=pe,
        bias=bias,
        window_size=window_size,
        block_num=block_num,
    )

    return ModelDescriptor(
        model, state_dict, architecture="OmniSR", scale=upscale, input_channels=num_in_ch, output_channels=num_out_ch
    )
