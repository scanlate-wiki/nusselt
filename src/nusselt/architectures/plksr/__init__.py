import math

from .plksr import PLKSR
from ...registry.model_descriptor import StateDict, ModelDescriptor
from ...utilities.state import get_seq_len


def load(state_dict: StateDict) -> ModelDescriptor[PLKSR]:
    state_keys = state_dict.keys()

    dim = 64
    n_blocks = 28
    # CCM options
    ccm_type = 'DCCM'
    # LK Options
    kernel_size = 17
    split_ratio = 0.5
    lk_type = 'PLK'
    # LK Rep options
    use_max_kernel = True
    sparse_kernels = [5, 5, 5, 5]
    sparse_dilations = [1, 2, 3, 4]
    with_idt = False  # not detect
    # EA ablation
    use_ea = False
    # Mobile Convert options
    is_coreml = False  # not detect
    scale = 4

    num_layers = get_seq_len(state_dict, "feats")
    scale = int((state_dict[f"feats.{num_layers - 1}.weight"].shape[0] / 3) ** 0.5)
    n_blocks = num_layers - 2
    ccm_shape = state_dict["feats.1.channe_mixer.2.weight"].shape[3]
    if ccm_shape == 3:
        ccm_type = "DCCM"
    elif ccm_shape == 1:
        ccm_type = "CCM"
    dim = state_dict["feats.0.weight"].shape[0]

    if "feats.1.lk.mn_conv.weight" in state_keys:
        lk_type = "RectSparsePLK"
        split_ratio = state_dict["feats.1.lk.mn_conv.weight"].shape[0] / dim
        kernel_size = state_dict["feats.1.lk.mn_conv.weight"].shape[3]
    elif "feats.1.lk.convs.0.weight" in state_keys:
        lk_type = "SparsePLK"
        split_ratio = state_dict["feats.1.lk.convs.0.weight"].shape[0] / dim
        len_sparse_kernels = get_seq_len(state_dict, f"feats.1.lk.convs")
        kernel_size = state_dict[f"feats.1.lk.convs.{len_sparse_kernels - 1}.weight"].shape[3]
        if kernel_size != 5:
            use_max_kernel = True
            len_sparse_kernels -= 1
        sparse_kernels = [5 for _ in range(len_sparse_kernels)]
        sparse_dilations = [i + 1 for i in range(len_sparse_kernels)]
    else:
        split_ratio = state_dict["feats.1.lk.conv.weight"].shape[0] / dim
        kernel_size = state_dict["feats.1.lk.conv.weight"].shape[3]

    if "feats.1.attn.f.0.weight" in state_keys:
        use_ea = True
    model = PLKSR(
        dim=dim,
        n_blocks=n_blocks,
        scale=scale,
        ccm_type=ccm_type,
        kernel_size=kernel_size,
        split_ratio=split_ratio,
        lk_type=lk_type,
        use_max_kernel=use_max_kernel,
        sparse_kernels=sparse_kernels,
        sparse_dilations=sparse_dilations,
        with_idt=with_idt,
        use_ea=use_ea,
        is_coreml=is_coreml

    )
    return ModelDescriptor(
        model,
        state_dict,
        architecture="PLKSR",
        scale=scale,
        input_channels=3,
        output_channels=3
    )
