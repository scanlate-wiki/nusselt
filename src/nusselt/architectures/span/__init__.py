from .span import SPAN
from ...registry.model_descriptor import StateDict, ModelDescriptor
from ...utilities.state import get_scale_and_output_channels


def load(state_dict: StateDict) -> ModelDescriptor[SPAN]:
    img_range = 255.0
    rgb_mean = (0.4488, 0.4371, 0.4040)

    num_in_ch = state_dict["conv_1.sk.weight"].shape[1]
    feature_channels = state_dict["conv_1.sk.weight"].shape[0]

    upscale, num_out_ch = get_scale_and_output_channels(
        state_dict["upsampler.0.weight"].shape[0],
        num_in_ch,
    )
    bias = "block_1.c1_r.sk.bias" in state_dict

    model = SPAN(
        num_in_ch,
        num_out_ch,
        feature_channels,
        upscale,
        bias,
        img_range,
        rgb_mean,
    )

    return ModelDescriptor(
        model,
        state_dict,
        architecture="SPAN",
        scale=upscale,
        input_channels=num_in_ch,
        output_channels=num_out_ch
    )
