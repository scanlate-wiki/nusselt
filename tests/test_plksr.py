from nusselt.architectures.plksr import PLKSR,load
from tests.util import (
    ModelFile,
    TestImage,
    assert_loads_correctly,
    assert_image_inference,
    disallowed_props,
)


def test_plksr_load():
    assert_loads_correctly(
        load,
        lambda: PLKSR(),
        lambda:PLKSR(n_blocks=12,kernel_size=13,use_ea=False),
        lambda: PLKSR(scale=1),
        lambda: PLKSR(scale=2),
        lambda: PLKSR(scale=3),
        lambda: PLKSR(scale=4),
    )


def test_compact_inference(snapshot):
    file = ModelFile(name="net_g_80000.pth")

    model = file.load_model()

    assert model == snapshot(exclude=disallowed_props)
    assert isinstance(model.model, PLKSR)

    assert_image_inference(
        file,
        model,
        [TestImage.COLOR_64],
    )
