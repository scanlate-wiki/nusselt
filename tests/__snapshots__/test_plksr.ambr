# serializer version: 1
# name: test_compact_inference
  ModelDescriptor(
    architecture='PLKSR',
    input_channels=3,
    output_channels=3,
    scale=4,
  )
# ---
